openpgp/packet/packet.go

```diff
@@ -525,6 +526,83 @@ func writeMPI(w io.Writer, bitLength uint16, mpiBytes []byte) (err error) {
 	return
 }
 
+func WritePaddedBigInt(w io.Writer, length int, X *big.Int) (n int, err error) {
+	bytes := X.Bytes()
+	n1, err := w.Write(make([]byte, length-len(bytes)))
+	if err != nil {
+		return n1, err
+	}
+	n2, err := w.Write(bytes)
+	if err != nil {
+		return n2, err
+	}
+	return (n1 + n2), err
+}
+
+// Minimum number of bytes to fit the curve coordinates. All
+// coordinates have to be 0-padded to this length.
+func mpiPointByteLength(curve elliptic.Curve) int {
+	return (curve.Params().P.BitLen() + 7) / 8
+}
+
+func readPointMPI(r io.Reader, curve elliptic.Curve) (X, Y *parsedMPI, err error) {
+	X = nil
+	Y = nil
+	var buf [2]byte
+	_, err = readFull(r, buf[0:2])
+	if err != nil {
+		return
+	}
+
+	//coordBitLen := mpiPointByteLength(curve) * 8
+	mpiBitSize := int(buf[0])<<8 | int(buf[1])
+	mpiByteSize := (mpiBitSize + 7) / 8
+	r.Read(buf[:1])
+	header := buf[0]
+	if header == 0x4 {
+		//if mpiBitSize != 2 * coordBitLen + 3 {
+		//	err = errors.StructuralError("Invalid MPI packet size.")
+		//	return
+		//}
+
+		coordLen := (mpiByteSize - 1) / 2
+
+		X = new(parsedMPI)
+		X.bytes = make([]byte, coordLen)
+		X.bitLength = uint16(coordLen * 8)
+		_, err = readFull(r, X.bytes)
+		if err != nil {
+			return
+		}
+
+		Y = new(parsedMPI)
+		Y.bytes = make([]byte, coordLen)
+		Y.bitLength = uint16(coordLen * 8)
+		_, err = readFull(r, Y.bytes)
+		if err != nil {
+			return nil, nil, err
+		}
+	} else if header == 0x40 {
+		//if mpiBitSize != coordBitLen + 7 {
```

Comment:
I think it's because when the point is read during EncryptedKey parsing, we don't have our curve yet because we've yet to find our ecdh PrivateKey for that EncryptedKey: https://github.com/zapu/go-crypto/blob/ecdh-rls2/openpgp/packet/encrypted_key.go#L53

They coord bit length could be checked afterwards somewhere, when the key and curve are known.